controllers/topology/util.go

```diff
@@ -167,3 +185,111 @@ func refToUnstructured(ref *corev1.ObjectReference) *unstructured.Unstructured {
+		return nil, errors.Wrapf(err, "failed to list MachineSets for MachineDeployment/%s", md.Name)
+	}
+
+	// Copy the MachineSets to an array of MachineSet pointers, to avoid MachineSet copying later.
+	res := make([]*clusterv1.MachineSet, 0, len(msList.Items))
+	for i := range msList.Items {
+		res = append(res, &msList.Items[i])
+	}
+	return res, nil
+}
+
+// calculateTemplatesInUse returns all templates referenced in non-deleting MachineDeployment and MachineSets.
+func calculateTemplatesInUse(md *clusterv1.MachineDeployment, msList []*clusterv1.MachineSet) (map[string]bool, error) {
+	templatesInUse := map[string]bool{}
+
+	// Templates of the MachineSet are still in use if the MachineSet is not in deleting state.
+	for _, ms := range msList {
+		if !ms.DeletionTimestamp.IsZero() {
+			continue
+		}
+
+		bootstrapRef := ms.Spec.Template.Spec.Bootstrap.ConfigRef
+		infrastructureRef := &ms.Spec.Template.Spec.InfrastructureRef
+		if err := addTemplateRef(templatesInUse, bootstrapRef, infrastructureRef); err != nil {
+			return nil, errors.Wrapf(err, "failed to add templates of %s to templatesInUse", KRef{Obj: ms})
+		}
+	}
+
+	// If MachineDeployment has already been deleted or still exists and is in deleting state, then there
+	// are no templates referenced in the MachineDeployment which are still in use, so let's return here.
+	if md == nil || !md.DeletionTimestamp.IsZero() {
+		return templatesInUse, nil
+	}
+
+	//  Otherwise, the templates of the MachineDeployment are still in use.
+	bootstrapRef := md.Spec.Template.Spec.Bootstrap.ConfigRef
+	infrastructureRef := &md.Spec.Template.Spec.InfrastructureRef
+	if err := addTemplateRef(templatesInUse, bootstrapRef, infrastructureRef); err != nil {
+		return nil, errors.Wrapf(err, "failed to add templates of %s to templatesInUse", KRef{Obj: md})
+	}
+	return templatesInUse, nil
+}
+
+// deleteTemplateIfUnused deletes the template (ref), if it is not in use (i.e. in templatesInUse).
+func deleteTemplateIfUnused(ctx context.Context, c client.Client, templatesInUse map[string]bool, ref *corev1.ObjectReference) error {
+	// If ref is nil, do nothing (this can happen, because bootstrap templates are optional).
+	if ref == nil {
+		return nil
+	}
+
+	log := loggerFrom(ctx).WithRef(ref)
+
+	refID, err := templateRefID(ref)
+	if err != nil {
+		return errors.Wrapf(err, "failed to calculate templateRefID")
+	}
+
+	// If the template is still in use, do nothing.
+	if templatesInUse[refID] {
+		// TODO(sbueringer) use KRef after 5153 has been merged
+		log.V(3).Infof("Not deleting %s/%s, because it's still in use", ref.Kind, ref.Name)
+		return nil
+	}
+
+	// TODO(sbueringer) use KRef after 5153 has been merged
+	log.Infof("Deleting %s/%s", ref.Kind, ref.Name)
+	if err := external.Delete(ctx, c, ref); err != nil && !apierrors.IsNotFound(err) {
+		// TODO(sbueringer) use KRef after 5153 has been merged
```

Comment:
Can you explain? I'm not sure what I'm missing. To me it looks like I'm already tolerating the isnotfound error
```go
if err := external.Delete(ctx, c, ref); err != nil && !apierrors.IsNotFound(err)
```

(the conversation shows outdated, probably because it has been moved to another place, but the IsNotFound has been there for a while :))